### DISCUSSION TITLE
Multiple Client Certificates Sharing Common Issuer and Subject

### DISCUSSION THREAD

**Discussion by ApoorvaShetty16** [openssl/openssl#22582] [Q&A] [answered]
Scenario:
Two clients have independently created self-signed certificates for SSL/TLS communication. These certificates share identical information in their "Subject" and "Issuer" fields. Both client certificates are getting added into the server's truststore. However, when attempting to establish an SSL connection, the server is able to accept first connection from client and other client connection attempt fails with 'self signed certificate error'.

Client-1 cert :

Certificate:
...
        Issuer: C=US, ST=California, L=Palo Alto, O=xyz, Inc./emailAddress=client@xyz.com, CN=abc
        Subject: C=US, ST=California, L=Palo Alto, O=xyz, Inc./emailAddress=ssl-certificates@xyz.com, CN=abc
...

Client-2 cert:

Certificate:
...
       Issuer: C=US, ST=California, L=Palo Alto, O=xyz, Inc./emailAddress=client@xyz.com, CN=abc
       Subject: C=US, ST=California, L=Palo Alto, O=xyz, Inc./emailAddress=ssl-certificates@xyz.com, CN=abc
...

Question:
-Is Same Issuer/Subject the reason for failure here? Is it mandatory to have clients with unique "Issuer" and "Subject" fields in this context ?

-What parameters will be considered while verifying the certificates?

**Comment by t8m** [comment, marked as answer]
Unfortunately in this scenario having multiple different self-signed certificates in the trust store will not work properly as one of the certificates will match the subject/issuer name matching first and the other won't be considered.

**Reply by ApoorvaShetty16** [reply]
Thanks @t8m . So is it expected here that the server rejects the other clients connection with "18-self signed" error?

**Reply by t8m** [reply]
Do the certificates have X509v3 Subject Key Identifier and X509v3 Authority Key Identifier set?

**Reply by ApoorvaShetty16** [reply]
We do not have any X509v3 extensions in our client certificates.

**Reply by t8m** [reply]
If X509v3 Subject Key Identifier and X509v3 Authority Key Identifier were set appropriately, it would probably make your use case work.

**Reply by ApoorvaShetty16** [reply]
Thanks for the suggestion @t8m . Can you also help answer below queries?

How can we differentiate the above error in this scenario with the error that we get due to the certificate not found in truststore? 
Also , Is there any OpenSSL method to identify this case on handshake failure or while adding certificates to truststore?

**Reply by t8m** [reply]
AFAIK there is no simple way for neither of these things. While adding certificates to trust store it would be programmatically possible to check whether there already is a certificate with the same subject but you would have to AFAIK write an application for that.
